R/print.data.table.R

```diff
@@ -164,3 +186,39 @@ shouldPrint = function(x) {
 #   as opposed to printing a blank line, for excluding col.names per PR #1483
 cut_top = function(x) cat(capture.output(x)[-1L], sep = '\n')
 
+# to calculate widths of data.table for PR #4074
+# gets the width of the data.table at each column
+#   and compares it to the console width
+dt_width = function(x, class, row.names, col.names) {
+  widths = apply(nchar(x, type='width'), 2L, max)
+  if (class) widths = ifelse(widths < 6L, 6L, widths)
+  if (col.names != "none") names = sapply(colnames(x), nchar, type = "width") else names = 0L
+  dt_widths = ifelse(widths > names, widths, names)
+  rownum_width = if (row.names) as.integer(ceiling(log10(nrow(x)))+1) else 0L
+  cumsum(dt_widths + 1L) + rownum_width + 1L
```

Comment:
another micro-optimization -- we can do `+2 else 0L`. I think `if(!row.names)` the `+1L` in the next step is wrong. However I also just noticed this:

https://github.com/Rdatatable/data.table/issues/4083

So even if `row.names=FALSE`, it could still occupy some characters. I think you can safely ignore that in this PR, rather than complicating logic to detect whether `---` will be present.

Reply:
I wondered about the `---` but hadn't done anything about it yet. Imagine we'll want to wait for that to be resolved before deciding any padding.

I believe the `+1L` in `cumsum(dt_widths + 1L)` should account for the space between the row names and the first column instead of needing to add another to the `rownum_width`.

Reply:
But I'll take off the `+1L` after the `rownum_width` bc I think you are right that it is unnecessary.

Reply:
I think the `+1L` is right when `row.names=TRUE`